### PR TITLE
Allow building godot with Clang on WIndows

### PR DESCRIPTION
**Pull request by Adriankhl**
Tested on 3.2-stable

Related to https://github.com/godotengine/godot/issues/6560

Now we can build godot on WIndows with clang and WIndows SDK. 

Also the `scons p=windows tools=yes use_mingw=yes use_llvm=yes bits=64 use_thinlto=yes target=release_debug -j8` in https://github.com/godotengine/godot/issues/11340 doesn't make sense since `use_thinlto` has to be used with `use_lto` previously. And the build failed when I enable them all. This PR enable `use_lto` automatically when `use_thinlto=yes`, and fix the failed build by using llvm tools instead of mingw tools.

mingw + llvm build: `scons platform=windows target=release_debug bits=64 use_mingw=yes use_llvm=yes use_thinlto=yes`

For some reasons, I have to use https://github.com/mstorsjo/llvm-mingw instead of the llvm packages in msys2 in order to compile the mingw + llvm build.

windowsSDK + llvm build: `scons platform=windows target=release_debug bits=64 use_llvm=yes use_thinlto=yes` 

I have also created a guide on using llvm and GIt Bash to compile Godot, see https://github.com/Adriankhl/llvm-git-bash . It may not be necessary to extract the environment variables from developer powershell since scons seems to be able to extract that automatically? I know cmake need that but I am not sure for scons.

### CODE REVIEW

File: platform/windows/detect.py [hunk 1]
```diff
@@ -216,6 +223,8 @@ def configure_msvc(env, manual_msvc_config):
     env.AppendUnique(CPPDEFINES=['NOMINMAX']) # disable bogus min/max WinDef.h macros
     if env["bits"] == "64":
         env.AppendUnique(CPPDEFINES=['_WIN64'])
+    if env['use_llvm']:
+        env.Append(CCFLAGS=['-mssse3', '-msse4.1', '-Wno-c++11-narrowing'])
```
Comment (akien-mga) [inline platform/windows/detect.py]: What's the reason for these changes? This will prevent running Godot on any platform that doesn't support SSE 4.1.

Comment (Adriankhl) [reply platform/windows/detect.py]: I need those flags to to build it successfully on my PC.

I should admit that I am not too familiar with how those instructions work. It seems like SSSE3 and SSE 4.1 features in those `thirdparty` libraries are enabled automatically, but clang-cl doesn't know that. Do you know how the availability of the instruction sets are detected in Windows?

`-Wno-c++11-narrowing` is set because clang-cl complains something from the Windows header file.

Comment (Adriankhl) [reply platform/windows/detect.py]: At least I can rerun the build without those flags and open an issue to document why those flags are added.

Comment (lawnjelly) [reply platform/windows/detect.py]: It may be for CPU detection and dynamic dispatch in the third party code. MSVC allows you to explicitly use e.g. intrinsics above the architecture specified in the cl.exe command line, whereas gcc (and probably clang) do not afaik. You have to take _great care_ around this area to prevent instructions being compiled in above the level of the CPU (which will cause crashes).

Presumably the non-msvc build already has to get around this issue (there may simply be defines to prevent these instructions being compiled at all in third party libraries).

Comment (Adriankhl) [reply platform/windows/detect.py]: I just realized that clang has a `-march=native` flag like gcc which may solve the problem. Let me do some testing. Python also provide some functionality to get cpu information.

Comment (lawnjelly) [reply platform/windows/detect.py]: > I just realized that clang has a `-march=native` flag like gcc which may solve the problem. Let me do some testing. Python also provide some functionality to get cpu information.

This may be confusing the architecture doing the compile and the target architecture. The build CPU is largely irrelevant (after all we e.g. compile for ARM on x86), the relevant thing here is the minimum architecture for the target platform.

Comment (akien-mga) [reply platform/windows/detect.py]: We don't pass any of those for Linux builds (`platform/x11/detect.py`), so it shouold not be necessary for Windows. If some thirdparty code fails on this, we need to investigate why, it means it's not doing its detection properly.

Comment (Adriankhl) [reply platform/windows/detect.py]: I see. Let's see if I can understand how those thirdparty libraries (such as `libwebp`) get the information. 

This is one of the error messages:
```
thirdparty\libwebp\src\dsp\alpha_processing_sse41.c(52,26): error: always_inline function '_mm_shuffle_epi8' requires target feature 'ssse3', but would be
      inlined into function 'ExtractAlpha_SSE41' that is compiled without support for 'ssse3'
      const __m128i b0 = _mm_shuffle_epi8(a0, kCstAlpha0);
                         ^
thirdparty\libwebp\src\dsp\alpha_processing_sse41.c(53,26): error: always_inline function '_mm_shuffle_epi8' requires target feature 'ssse3', but would be
      inlined into function 'ExtractAlpha_SSE41' that is compiled without support for 'ssse3'
      const __m128i b1 = _mm_shuffle_epi8(a1, kCstAlpha1);
                         ^
thirdparty\libwebp\src\dsp\alpha_processing_sse41.c(54,26): error: always_inline function '_mm_shuffle_epi8' requires target feature 'ssse3', but would be
      inlined into function 'ExtractAlpha_SSE41' that is compiled without support for 'ssse3'
      const __m128i b2 = _mm_shuffle_epi8(a2, kCstAlpha2);
                         ^
thirdparty\libwebp\src\dsp\alpha_processing_sse41.c(55,26): error: always_inline function '_mm_shuffle_epi8' requires target feature 'ssse3', but would be
      inlined into function 'ExtractAlpha_SSE41' that is compiled without support for 'ssse3'
      const __m128i b3 = _mm_shuffle_epi8(a3, kCstAlpha3);
```

Comment (akien-mga) [reply platform/windows/detect.py]: The relevant logic is in `dsp/dsp.h`:
```
#if defined(__SSE2__) || defined(WEBP_MSC_SSE2) || defined(WEBP_HAVE_SSE2)
#define WEBP_USE_SSE2
#endif

#if defined(__SSE4_1__) || defined(WEBP_MSC_SSE41) || defined(WEBP_HAVE_SSE41)
#define WEBP_USE_SSE41
#endif
```

I assume you're having this issue when using Clang with MSVC?
libwebp is written with the assumption that MSVC means that it can do what @lawnjelly described in https://github.com/godotengine/godot/pull/36580#discussion_r385167821, but evidently that's not possible when using Clang instead of MS VC++.

**Comment by akien-mga** [comment]
Closing as there are outstanding issues which haven't been addressed yet.

There's still a lot of relevant work done here so it can be used as a base for a new PR implementing support for using `clang-cl` on Windows.
